### PR TITLE
progress bar: use spinners for -1 blob sizes

### PR DESCRIPTION
**Pull request by vrothberg**
Schema1 manifests don't include blob sizes causing the BlobInfo sizes to
be -1 and ultimately preventing the progress bars from displaying
meaningful data.  Use a spinner for such cases.  The spinner renders a
sequence of dots (".", "..", "...", "....", "").

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
@rhatdan @mtrmac @edsantiago PTAL

You can easily test it with `go mod edit -replace github.com/containers/image=github.com/vrothberg/image@use-spinners` and revendoring in Skopeo or Podman.

File: copy/copy.go [hunk 1]
```diff
@@ -597,15 +597,32 @@ func (c *copier) createProgressBar(pool *mpb.Progress, info types.BlobInfo, kind
...
-			decor.Name(prefix),
-		),
-		mpb.AppendDecorators(
-			decor.OnComplete(decor.CountersKibiByte("%.1f / %.1f"), " "+onComplete),
-		),
-	)
+	// If we don't know the size of the blob (size == -1), just display a
+	// spinner. Otherwise, display a proper progress bar.
+	var bar *mpb.Bar
+	if info.Size == -1 {
```
Comment (edsantiago) [inline copy/copy.go]: Is there a guarantee that the only invalid size will be `-1`, and never zero or -2 or anything else?

And, minor point, I tend to like coding for the usual-case first (vs rare-case). What would you think of:
```go
    if info.Size > 0 {
        bar = pool.AddBar(...)
    } else {
        // Unknown or invalid size
        bar = pool.AddSpinner()
    }
```

Comment (TomSweeneyRedHat) [reply copy/copy.go]: +1 to the `<0' vs `==-1' thought form @edsantiago

File: vendor.conf [hunk 2]
```diff
@@ -46,6 +46,6 @@ github.com/etcd-io/bbolt v1.3.2
 github.com/klauspost/pgzip v1.2.1
 github.com/klauspost/compress v1.4.1
 github.com/klauspost/cpuid v1.2.0
-github.com/vbauerster/mpb v3.3.4
+github.com/vbauerster/mpb v3.4.0
```
Comment (TomSweeneyRedHat) [inline vendor.conf]: Is this change intentional?

File: copy/copy.go (hunk 1 above)
Comment (mtrmac) [reply copy/copy.go]: The definition of `BlobInfo.Size` explicitly says `-1`.

(The flip side for having the usual case first is that one then has to explicitly track (or, as in the above suggestion, comment) what the unusual case is. In this case I don’t care either way at all, neither about the ordering nor about the condition)

File: copy/copy.go [hunk 3]
```diff
@@ -597,15 +597,32 @@ func (c *copier) createProgressBar(pool *mpb.Progress, info types.BlobInfo, kind
...
+	} else {
+		bar = pool.AddBar(info.Size,
+			mpb.BarClearOnComplete(),
+			mpb.PrependDecorators(
+				decor.Name(prefix),
+			),
+			mpb.AppendDecorators(
+				decor.OnComplete(decor.CountersKibiByte("%.1f / %.1f"), " "+onComplete),
+			),
+		)
```
Comment (mtrmac) [inline copy/copy.go]: It seems attractive to build the shared `[]mpb.BarOption` options outside of the `if`/`else` — but it might well be so much uglier than the current code that it would not be worth it. Up to you.

File: vendor.conf (hunk 2 above)
Comment (mtrmac) [reply vendor.conf]: Yes, that version adds the `AddSpinner` call.

File: copy/copy.go (hunk 1 above)
Comment (vrothberg) [reply copy/copy.go]: 2 pro and 1 neutral, so I'll change it as proposed.

File: copy/copy.go (hunk 3 above)
Comment (vrothberg) [reply copy/copy.go]: Let's keep it as is but I see wisdom in rethinking this code when more conditions are being added.

**Comment by vrothberg** [comment]
Here's the PR for Skopeo: https://github.com/containers/skopeo/pull/689

**Comment by vrothberg** [comment]
Note that this fix was motivated by https://bugzilla.redhat.com/show_bug.cgi?id=1695236.

**Comment by vrothberg** [comment]
Skopeo PR is green. Force merging here, to revendor into Skopeo.

File: copy/copy.go [hunk 4]
```diff
@@ -597,15 +597,32 @@ func (c *copier) createProgressBar(pool *mpb.Progress, info types.BlobInfo, kind
...
-			decor.Name(prefix),
-		),
-		mpb.AppendDecorators(
-			decor.OnComplete(decor.CountersKibiByte("%.1f / %.1f"), " "+onComplete),
-		),
-	)
+	// Use a normal progress bar when we know the size (i.e., size > 0).
+	// Otherwise, use a spinner to indicate that something's happening.
+	var bar *mpb.Bar
+	if info.Size > 0 {
```
Comment (mtrmac) [inline copy/copy.go]: Pedantic nitpick, just to get it off my chest: `≥0`. Not really worth fixing.
